internal/assertion/update.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2020 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package assertion
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"time"
+
+	"github.com/fatih/color"
+	"k8s.io/klog"
+
+	"sigs.k8s.io/krew/internal/version"
+)
+
+const (
+	githubVersionURL = "https://api.github.com/repos/kubernetes-sigs/krew/releases/latest"
+
+	upgradeNotification = "You are using an old version of krew (%s). Please upgrade to the latest version (%s)"
+)
+
+// for testing
+var versionURL = githubVersionURL
+
+// CheckVersion returns a notification message to inform the user
+// about a new version of krew, or an empty string. The notification
+// will only be emitted once a day.
+func CheckVersion(basePath string) string {
+	f := filepath.Join(basePath, "last_update_check")
+	lastCheck := loadTimestamp(f)
+	tag := fetchTag(lastCheck)
+	if version.GitTag() == tag {
+		return ""
+	}
+	saveTimestamp(f)
+	return color.New(color.Bold).Sprintf(upgradeNotification, version.GitTag(), tag)
+}
+
+// fetchTag tries to return the tag name of the most recent krew
+// release. If the last check happened recently, or an error occurs
+// the hardcoded tag of the executing krew binary will be returned.
+// This effectively disables the update notification message.
+func fetchTag(lastCheck time.Time) string {
+	if time.Since(lastCheck).Hours() <= 24 {
+		klog.V(3).Info("Last check was recently, skipping update check")
+		return version.GitTag()
+	}
+
+	latestTag, err := fetchLatestTagFromGithub()
+	if err != nil || latestTag == "" {
+		klog.V(3).Infof("Could not fetch most recent tag name: %s", err)
```

Comment:
ditto: wrap; not log.

Reply:
ok, got it :)
However, in the outermost layer, I'll simply log the error instead of passing it onwards. The success of the main action should not depend on whether the update check succeeds or fails.